repose-aggregator/commons/utilities/src/main/scala/org/openrepose/commons/utils/servlet/http/HttpServletRequestWrapper.scala

```diff
@@ -0,0 +1,150 @@
+  extends javax.servlet.http.HttpServletRequestWrapper(originalRequest)
+  with HeaderInteractor {
+
+  object RequestBodyStatus extends Enumeration {
+    val Available, InputStream, Reader = Value
+  }
+
+  private var status = RequestBodyStatus.Available
+
+  def this(originalRequest: HttpServletRequest) = this(originalRequest, originalRequest.getInputStream())
+
+  val caseInsensitiveOrdering = Ordering.by[String, String](_.toLowerCase)
+
+  private var headerMap: Map[String, List[String]] = new TreeMap[String, List[String]]()(caseInsensitiveOrdering)
+  private var removedHeaders: Set[String] = new TreeSet[String]()(caseInsensitiveOrdering)
+
+  override def getInputStream: ServletInputStream = {
+    if (status == RequestBodyStatus.Reader) throw new IllegalStateException else status = RequestBodyStatus.InputStream
+    inputStream
+  }
+
+  override def getReader: BufferedReader = {
+    if (status == RequestBodyStatus.InputStream) throw new IllegalStateException else status = RequestBodyStatus.Reader
+    new BufferedReader(new InputStreamReader(inputStream))
+  }
+
+  def getHeaderNamesScala: Set[String] = headerMap.keySet ++ super.getHeaderNames.asScala.toSet.filterNot(removedHeaders.contains)
+
+  override def getHeaderNames: util.Enumeration[String] = getHeaderNamesScala.toIterator.asJavaEnumeration
+
+  override def getHeaderNamesList: util.List[String] = getHeaderNamesScala.toList.asJava
+
+  override def getIntHeader(headerName: String): Int = Option(getHeader(headerName)).getOrElse("-1").toInt
+
+  def getHeadersScala(headerName: String): List[String] = {
+    if (removedHeaders.contains(headerName)) {
+      List[String]()
+    }
+    else {
+      headerMap.getOrElse(headerName, super.getHeaders(headerName).asScala.toList)
+    }
+  }
+
+  override def getHeaders(headerName: String): util.Enumeration[String] = getHeadersScala(headerName).toIterator.asJavaEnumeration
+
+  override def getDateHeader(headerName: String): Long = Option(getHeader(headerName)).map(DateUtils.parseDate(_).getTime).getOrElse(-1)
+
+  override def getHeader(headerName: String): String = getHeadersScala(headerName).headOption.orNull
+
+  override def getHeadersList(headerName: String): util.List[String] = getHeadersScala(headerName).asJava
+
+  override def addHeader(headerName: String, headerValue: String): Unit = {
+    val existingHeaders: List[String] = getHeadersScala(headerName) //this has to be done before we remove from the list,
+                                                                    // because getting this list is partially based on the contents of the removed list
+    if (removedHeaders.contains(headerName)) {
+      removedHeaders = removedHeaders - headerName
+    }
+    headerMap = headerMap + (headerName -> (existingHeaders :+ headerValue))
+  }
+
+  override def addHeader(headerName: String, headerValue: String, quality: Double): Unit = addHeader(headerName, headerValue + ";q=" + quality)
```

Comment:
So the case I'm thinking of is when a user calls getHeader or getHeaders and gets back a value with a quality in it.

Imagine you have a request with the following headers:
x-auth-header: panda;q=0.7
x-keystone-header: rat;q=0.7

And you want to modify your request so that those headers are combined like so:
x-auth-header: panda;q=0.7, rat;q=0.8

Is there a way to achieve that with the current code without having to parse the value outside of the wrapper? What I would try and do is get the value of x-keystone-header, append it to x-auth-header with a new quality, then remove x-keystone-header.